File: 2021/02/02.py
def generator(input):
    commands = []
    for command in input.splitlines():
        dir, steps = command.split(' ')
        match dir:
            case 'forward': commands.append((int(steps), 0))
            case 'up': commands.append((0, - int(steps)))
            case 'down': commands.append((0, int(steps)))
    return commands

def part_1(input): 
    return solver(input, False)

def part_2(input): 
    return solver(input, True)

def solver(commands, track_aim):
    x, y, aim = 0, 0, 0
    for dx, dy in commands:
        x += dx
        aim += dy
        y += [dy, dx * aim][track_aim]
    return x * y

File: 2021/02/02.rs
aoc2021::main!();

fn generator(input: &str) -> Vec<(i32, i32)> {
    input
    .lines()
    .map(|command| command
        .splitn(2,' ')
        .collect_tuple()
        .unwrap())
    .map(|(dir, steps)| match dir {
        "forward"   => (steps.parse().unwrap(), 0),
        "down"      => (0, steps.parse().unwrap()),
        "up"        => (0, -steps.parse::<i32>().unwrap()),
        _ => unreachable!()})
    .collect_vec()   
}

fn part_1(input: Vec<(i32, i32)>) -> i32 {
    solver(input, false)
}

fn part_2(input: Vec<(i32, i32)>) -> i32 {
    solver(input, true)
}

fn solver(commands: Vec<(i32, i32)>, track_aim: bool) -> i32 {
    let (a, b, _) = commands
    .iter()
    .fold((0,0,0), |(x, y, aim), (dx, dy)| {
    if track_aim {(x + dx, y + dx * aim, aim + dy)}
    else {(x + dx, y + dy, 0)}});
    a * b
}